packages/common/src/request.ts

```diff
@@ -1,43 +1,114 @@
 import Message from "./message";
-import Response from "./response";
-import { generateRandomNumber } from "./utils";
-
-const SEPERATOR = "|";
+import {
+  generateRandomNumber,
+  joinPartsToBody,
+  splitBodyToParts,
+  Identity,
+} from "./utils";
+import {
+  Envelope,
+  box_request,
+  unbox_request,
+  Session,
+} from "rpch-crypto/nodejs";
+import { utils } from "ethers";
 
 /**
  * Represents a request made by the RPCh.
  * To be send over the HOPR network via Request.toMessage().
  */
 export default class Request {
-  constructor(
+  private constructor(
     public readonly id: number,
-    public readonly origin: string,
     public readonly provider: string,
-    public readonly body: string
+    public readonly body: string,
+    public readonly entryNode: Identity,
+    public readonly exitNode: Identity,
+    public readonly session: Session
   ) {}
 
-  public static fromData(
-    origin: string,
+  /**
+   * Create a new Request
+   * @param provider
+   * @param body
+   * @param entryNode
+   * @param exitNode
+   * @returns Request
+   */
+  public static createRequest(
     provider: string,
-    body: string
+    body: string,
+    entryNode: Identity,
+    exitNode: Identity
   ): Request {
-    return new Request(generateRandomNumber(), origin, provider, body);
+    const id = generateRandomNumber();
+    const payload = joinPartsToBody(["request", provider, body]);
+    const envelope = new Envelope(
+      utils.toUtf8Bytes(payload),
+      entryNode.peerId.toB58String(),
+      exitNode.peerId.toB58String()
+    );
+    const session = box_request(envelope, exitNode.cryptoIdentity);
+
+    return new Request(id, provider, body, entryNode, exitNode, session);
   }
 
-  public static fromMessage(message: Message): Request {
-    const [type, origin, provider, ...body] = message.body.split(SEPERATOR);
+  /**
+   * Recreate a Request from an incoming Message
+   * @param message
+   * @param exitNode
+   * @returns Request
+   */
+  public static fromMessage(
+    message: Message,
+    exitNode: Identity,
+    lastRequestFromClient: bigint,
+    updateLastRequestFromClient: (counter: bigint) => any
```

Comment:
The arguments of the `updateLastRequestFromClient` must also contain the Entry node peer ID, because that is the key you use to store the last TS in the DB on the Exit node.
Similarly, on the Client you use the Exit node peer ID, so the same comment would apply for a lambda used in `Response.fromMessage` method.